### ISSUE TITLE
Intended use(s) of mesh primitives

### ISSUE THREAD

**Issue by donmccurdy** [KhronosGroup/glTF#1278]
For the three.js glTF exporter, @takahirox has proposed a change that allows us to export multi-material meshes efficiently: https://github.com/mrdoob/three.js/pull/13536. It raised some questions for me about how to load primitives in three.js, because three are more possible use cases than I had thought of:

1. Each primitive points to different attribute accessors, but share a material. Use case mentioned in the spec would be to "limit number of indices per draw call", and engines might very reasonably choose to merge them into one draw call.
2. Each primitive points to different attribute accessors, and different materials.
3. Each primitive points to the same attribute accessors, but different indices, and all use a single material. Again the purpose is probably to limit number of indices per draw call, but engines wishing to merge draw calls should concatenate indices rather than vertex attributes.
4. Each primitive points to the same attribute accessors, but different indices, and different materials.
5. A single mesh arbitrarily combines all of the above.
6. Multiple meshes, each with 1 primitive, use the same attribute accessors to represent attributes of a single mesh with different materials.

I'd only been thinking of cases (1), (2), and (6) and was pretty close to writing code that would have been very inefficient with (3) and (4), so thought I'd open an issue here. 

Anyone else have loaders or exporters making particular assumptions about this question? I don't necessarily think any spec change is needed, just wanted to get a better idea of what to expect and how to be consistent here.

Somewhat related to https://github.com/KhronosGroup/glTF/issues/1249.

**Comment by mlimper** [comment]
> limit number of indices per draw call

This implementation hint in the spec is a bit misleading. Actually, I can't imagine why one would want to strictly limit the number of *indices* in a buffer (meaning, the number of primitives in a draw call).
For 16 bit indices, one limits the number of indexed vertices to 65,536, and hence the range of indices to [0, 63535], and I would guess that this is what this comment wants to say... could suggest a reworded version?

> Anyone else have loaders or exporters making particular assumptions about this question?

Our exporter currently assumes that the client may only support 16 bit indices, so we batch stuff together by material (to achieve a low number of draw calls, do vertex cache optimization, ...) and then split again if necessary to have at maximum 65,536 vertices per draw.
I would guess that, in most situations, compared to 32 bit indices, this may not really be a serious performance caveat - there may even be situations where insisting on 32 bit indices makes things worse, for example when rendering a large scene and considering the trade-off between batch size (larger is better) and the granularity that can be used for culling (smaller batches -> better granularity for culling) and memory usage (half the size for 16 bit index data, also nice for transmission...). So, I don't think that the 16 bit index limit will be the core of the performance problem in most cases, or that it would be the most relevant issue to focus on.

I know that professional tools like Simplygon offered something called "Draw Call Reduction", and something like that must exist in any game pipeline. Just to given an example what I have in mind how things are usually done in practice (game people feel free to correct me):

1. An artists models a scene, assigning materials to objects and not caring much about how those objects should be grouped for *rendering*, but rather grouping them in just the right way for efficient *editing*. When you'd export that scene (consisting of, say 20K triangles) as-is, you may end up with 23 different textures, assigned to 42 different objects

2. A cool content optimization tool turns the unoptimized scene into an optimized asset with a texture atlas, needing just a single draw call and having cache-optimized vertex layout, etc.

3. The engine loads and renders the optimized asset - no significant further optimizations performed during runtime

Now, I can imagine that there are currently many experimental workflows where the second step - optimizing the asset for efficient rendering - is not performed at all, and instead an unoptimized scene is just exported to glTF / ThreeJS right away. And then the biggest issue will be switching textures / materials all the time, many draw calls for just a few vertices, etc. - is that the kind of situation you are referring to with (3) and (4)? If yes, these kind of optimizations should probably be performed by exporters where possible (for example, you probably don't want to create a texture atlas on the client side), and maybe a guide for best practices could help here.

It would probably be cool to illustrate best practices for a couple of different scenarios, describing suggested ways to handle primitives in glTF exporters and engines, and including practical tips, such as batching stuff together in the exporter, optimizing vertices for a more cache-friendly layout, creating texture atlases, and so on.

TL;DR: Optimizing 3D content for efficient rendering should be mostly the responsibility of the exporter - even more if we assume glTF should be "ready to render". Illustrating best practices by example could help to work into that direction.

**Comment by donmccurdy** [comment]
> Now, I can imagine that there are currently many experimental workflows where the second step - optimizing the asset for efficient rendering - is not performed at all, and instead an unoptimized scene is just exported to glTF / ThreeJS right away. And then the biggest issue will be switching textures / materials all the time, many draw calls for just a few vertices, etc. - is that the kind of situation you are referring to with (3) and (4)? If yes, these kind of optimizations should probably be performed by exporters where possible (for example, you probably don't want to create a texture atlas on the client side), and maybe a guide for best practices could help here.

This is the primary case I'm concerned with for three.js and A-Frame, yes. Currently I'm not aware of any tools at all for performing draw call reduction glTF models, so I've considered merging meshes with shared materials (but not creating texture atlases) at runtime when possible.

Great points, thanks. A lot of the guesswork in cases 1-6 would go away if we assume exporters will do optimization (and promote tools to do so).

**Comment by mlimper** [comment]
> Currently I'm not aware of any tools at all for performing draw call reduction glTF models

That sounds like a really cool open source project ("unoptimized glTF in, optimized glTF out"), doesn't it?
Could maybe even be a nice proposal to launch something like this as part of University coursework?(@pjcozzi ? :-) )

**Comment by donmccurdy** [comment]
I just recently learned from @bghgary that Simplygon has (at least experimental) glTF support: https://blogs.windows.com/buildingapps/2017/12/07/microsoft-announces-simplygon-cloud-optimizes-mixed-reality-development/

Have only had time so far to do a single quick test. It successfully simplified a 100mb model into a 1.5mb model (which was awesome, but maybe a little _too_ simple haha). I think my materials didn't quite come through right but didn't look into that, or try to find any knobs for adjustment really.

**Comment by ousttrue** [comment]
I think it is better to use mesh primitives for `submesh`.
`submesh` assumes the following call, for example:

```c
// pseudocode
setVBOandIBO (mesh);
var offset = 0;
foreach (var submesh: submeshes)
{
    setMaterial (submesh.material);
    gl.drawElements (sbumesh.mode, submesh.count, mesh.indicesType, offset);
    offset + = submesh.count;
}
```

submesh is a range of indices (offset, count), with material as a specific parameter.

Relatedly, in many apps, Blender and Unity, MorphTarget is per mesh, not per submesh. It is more convenient if targets are on mesh.

In order to use the mesh primitive for `submesh`,
Suggest moving primitive.indices, primitive.attributes, primitive.targets to mesh.
Also add offset and count to the primitive.
Also, change node.mesh to node.meshes.

With this, I think that six patterns can be expressed.

1. node.meshes has multi meshes. Each mesh has single primitives
2. node.meshes has multi meshes. Each mesh has single primitives
3. node.meshes has single mesh. mesh has multi primitives
4. node.meshes has single mesh. mesh has multi primitives
5. node.meshes has single mesh. mesh has multi primitives
6. node.meshes has multi meshes. Each mesh has single primitives?
